components/director/pkg/authenticator/key_iterator.go

```diff
@@ -0,0 +1,45 @@
+package authenticator
+
+import (
+	"github.com/kyma-incubator/compass/components/director/pkg/apperrors"
+	"github.com/lestrrat-go/jwx/jwk"
+)
+
+/*
+ * Copyright 2020 The Compass Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+type KeyIterator struct {
```

Comment:
Let's make this KeyIterator more generic and reuse it in `runtimemapping/token_verifier.go`

What I would propose is to name it something like `jwkKeyIterator` since in Visit func you expect jwk.Key.

Also you can add two fields as in the `runtimemapping/token_verifier.go` but make them funcs:

```go
type KeyIterator struct {
	AlgorithmCriteria func(string) bool
	IDCriteria        func(string) bool
	ResultingKey      interface{}
}
```

Then the if in Visit func would be something like:

```go
if keyIterator.AlgorithmCriteria(key.Algorithm()) && keyIterator.IDCriteria(key.KeyID()) {
		...
}
```

Once you have that in `runtimemapping/token_verifier.go` you can create an itterator which will look like this:

```go
keyIterator := &authenticator.KeyIterator{
		AlgorithmCriteria: func(alg string) bool {
			return token.Method.Alg() == alg
		},
		IDCriteria: func(id string) bool {
			return id == keyID
		},
	}
```

and in the middleware:

```go
keyIterator := &authenticator.KeyIterator{
				AlgorithmCriteria: func(alg string) bool {
					return token.Method.Alg() == alg
				},
				IDCriteria: func(id string) bool {
					return true
				},
			}
```

What do you think?